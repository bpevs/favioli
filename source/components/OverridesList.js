import React from "react"
import Override from "./Override"

export default class OverridesList extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      overrides: props.overrides,
    }
  }

  onChange(override) {
    const { emoji, filter, index, toDelete } = override
    const overrides = this.state.overrides

    if (toDelete)
      overrides.splice(index, 1)
    else {
      overrides[index] = overrides[index] || {}
      if (emoji != null) overrides[index].emoji = emoji
      if (filter != null) overrides[index].filter = filter
    }

    this.setState(
      { overrides },
      () => this.props.onChange(overrides)
    )
  }

  render() {
    return <div className="override-inputs">{
      this.state.overrides
        .map((override, index) => <Override
          key={index}
          index={index}
          autoFocus={index === this.state.overrides.length - 1}
          onChange={this.onChange.bind(this)}
          emoji={override.emoji}
          filter={override.filter}
        />)
        .concat(<Override
          key="new"
          canDelete={false}
          index={this.state.overrides.length}
          onChange={this.onChange.bind(this)}
        />)
    }
  </div>;
  }
}

OverridesList.defaultProps = {
  overrides: [],
}
