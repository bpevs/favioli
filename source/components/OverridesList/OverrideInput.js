import { Picker } from "emoji-mart"
import React from "react"
import { isRegexString } from "../../utilities/isRegexString/isRegexString"


const DEFAULT_EMOJI = {
  colons: ":grinning:",
  emoticons: [],
  id: "grinning",
  name: "Grinning Face",
  native: "ðŸ˜€",
  skin: null,
  unified: "1f600",
}

const DEFAULT_FILTER = ""


export class OverrideInput extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      pickerIsOpen: false,
    }
  }

  _changeFilter(evt) {
    this.props.onChange({
      filter: evt.target.value,
      index: this.props.index,
    })
    this.setState({
      pickerIsOpen: false,
    })
  }

  _delete() {
    this.props.onChange({
      index: this.props.index,
      toDelete: true,
    })
  }

  _selectEmoji(emoji) {
    this.setState(
      { pickerIsOpen: false },
      () => this.props.onChange({
        emoji,
        index: this.props.index,
      }),
    )
  }

  _togglePicker() {
    const pickerIsOpen = !this.state.pickerIsOpen
    this.setState({ pickerIsOpen })
  }

  render() {
    const { pickerIsOpen } = this.state
    const { emoji, filter } = this.props
    const filterColor = isRegexString(filter) ? "green" : "black"

    return <div className="override">
      <input
        autoFocus={this.props.autoFocus}
        className="filter"
        style={{ color: filterColor }}
        value={filter}
        onChange={this._changeFilter.bind(this)}
        placeholder="favioli.com" />

      <button
        children={emoji.native}
        className="emoji"
        onClick={this._togglePicker.bind(this)}
      />

      {
        this.props.canDelete
         ? <button className="remove" onClick={this._delete.bind(this)}>X</button>
         : ""
      }

      { pickerIsOpen ? <Picker
        style={{
          boxShadow: "5px 3px 20px rgba(0,0,0,0.2)",
          position: "absolute",
          right: 0,
          top: 0,
          transform: "translateY(52%) translateX(-30%)",
          zIndex: 10,
        }}
        emoji={emoji.id}
        native={true}
        onSelect={this._selectEmoji.bind(this)}
        showSkinTones={false}
        skin={1}
        title="Select Emoji"
      /> : ""}
    </div>
  }
}

OverrideInput.defaultProps = {
  canDelete: true,
  emoji: DEFAULT_EMOJI,
  filter: DEFAULT_FILTER,
  onChange: () => { return },
}
