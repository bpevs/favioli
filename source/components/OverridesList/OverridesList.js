import React from "react"
import { OverrideInput } from "./OverrideInput"

export class OverridesList extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      overrides: props.overrides,
    }
  }

  componentDidUpdate() {
    if (this.state.overrides !== this.props.overrides) {
      this.setState({ overrides: this.props.overrides })
    }
  }

  async onChange(override) {
    const { emoji, filter, index, toDelete } = override
    const overrides = this.state.overrides

    if (toDelete)
      overrides.splice(index, 1)
    else {
      overrides[index] = overrides[index] || {}
      if (emoji != null) overrides[index].emoji = emoji
      if (filter != null) overrides[index].filter = filter
    }

    this.setState({ overrides }, () => {
      this.props.onChange({ overrides })
    })
  }

  render() {
    return <div className="override-inputs">{
      this.state.overrides
        .map((override, index) => <OverrideInput
          key={index}
          index={index}
          autoFocus={index === this.state.overrides.length - 1}
          onChange={this.onChange.bind(this)}
          emoji={override.emoji}
          filter={override.filter}
        />)
        .concat(<OverrideInput
          key="new"
          canDelete={false}
          index={this.state.overrides.length}
          onChange={this.onChange.bind(this)}
        />)
    }
  </div>
  }
}

OverridesList.defaultProps = {
  onChange: () => console.log("No onChange method supplied"),
  overrides: [],
}
