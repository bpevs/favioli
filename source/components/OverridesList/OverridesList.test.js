import { mount, shallow } from "enzyme"
import * as React from "react"
import { OverridesList } from "./OverridesList"
import { Picker } from "emoji-mart"

const mockOverride = {
  emoji: {
      "colons": ":grinning:",
      "emoticons": [],
      "id": "grinning",
      "name": "Grinning Face",
      "native": "ðŸ˜€",
      "skin": null,
      "unified": "1f600",
  },
  filter: "bookface",
}

test("Should Render OverridesList", () => {
  const wrapper = shallow(<OverridesList
    onChange={jest.fn()}
  />)

  expect(wrapper).toMatchSnapshot()
})

test("Should Render OverridesList with existing overrides", () => {
  const wrapper = shallow(<OverridesList
    overrides={[ mockOverride ]}
    onChange={jest.fn()}
  />)

  expect(wrapper).toMatchSnapshot()
})

test("Should update with new overrides", () => {
  const wrapper = mount(<OverridesList
    overrides={[ mockOverride ]}
    onChange={jest.fn()}
  />)

  expect(wrapper).toMatchSnapshot()

  const mockOverride1 = Object.assign({}, mockOverride, { filter: "/bookface/" })
  wrapper.setProps({ "overrides": [ mockOverride, mockOverride1 ] })

  expect(wrapper).toMatchSnapshot()
})


test("Should edit filter", () => {
  const onChange = jest.fn()
  const wrapper = mount(<OverridesList
    overrides={[ mockOverride ]}
    onChange={onChange}
  />)
  expect(wrapper.state("overrides").length).toBe(1)
  expect(wrapper.find(".override-inputs").children().length).toBe(2)

  const override = wrapper.find(".override-inputs input").first()
  override.simulate("change", mockOverride)

  expect(onChange).toBeCalled()
  expect(wrapper.state("overrides").length).toBe(1)
  expect(wrapper.find(".override-inputs").children().length).toBe(2)
})


test("Should add override on editing last in list", () => {
  const onChange = jest.fn()
  const wrapper = mount(<OverridesList
    overrides={[ mockOverride ]}
    onChange={onChange}
  />)
  expect(wrapper.state("overrides").length).toBe(1)
  expect(wrapper.find(".override-inputs").children().length).toBe(2)

  const override = wrapper.find(".override-inputs input").last()
  override.simulate("change", mockOverride)

  expect(onChange).toBeCalled()
  expect(wrapper.state("overrides").length).toBe(2)
  expect(wrapper.find(".override-inputs").children().length).toBe(3)
})


test("Should delete override", () => {
  const onChange = jest.fn()
  const wrapper = mount(<OverridesList
    overrides={[ mockOverride ]}
    onChange={onChange}
  />)

  expect(wrapper.state("overrides").length).toBe(1)
  expect(wrapper.find(".override-inputs").children().length).toBe(2)

  const override = wrapper.find(".override-inputs button.remove").first()
  override.simulate("click")

  expect(onChange).toBeCalled()
  expect(wrapper.state("overrides").length).toBe(0)
  expect(wrapper.find(".override-inputs").children().length).toBe(1)
})


test("Should open emojimart on clicking emoji button", () => {
  const onChange = jest.fn()
  const wrapper = mount(<OverridesList
    onChange={onChange}
  />)

  const override = wrapper.find(".override-inputs button.emoji").first()
  override.simulate("click")
  expect(onChange).toHaveBeenCalledTimes(0)

  expect(wrapper.find(Picker).first().exists()).toBe(true)
})

test("Should run onChange and close emojimart on selecting emoji", () => {
  const onChange = jest.fn()
  const wrapper = mount(<OverridesList
    onChange={onChange}
  />)

  const override = wrapper.find(".override-inputs button.emoji").first()
  override.simulate("click", mockOverride)

  const emoji = wrapper.find("span.emoji-mart-emoji span").first()
  emoji.simulate("click")

  expect(onChange).toHaveBeenCalledTimes(1)
  expect(wrapper.find(Picker).first().exists()).toBe(false)
})
