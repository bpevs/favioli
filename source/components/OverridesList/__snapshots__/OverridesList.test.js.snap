// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should Render OverridesList 1`] = `
<div
  className="override-inputs"
>
  <OverrideInput
    canDelete={false}
    emoji={
      Object {
        "colons": ":grinning:",
        "emoticons": Array [],
        "id": "grinning",
        "name": "Grinning Face",
        "native": "ðŸ˜€",
        "skin": null,
        "unified": "1f600",
      }
    }
    filter=""
    index={0}
    key="new"
    onChange={[Function]}
  />
</div>
`;

exports[`Should Render OverridesList with existing overrides 1`] = `
<div
  className="override-inputs"
>
  <OverrideInput
    autoFocus={true}
    canDelete={true}
    emoji={
      Object {
        "colons": ":grinning:",
        "emoticons": Array [],
        "id": "grinning",
        "name": "Grinning Face",
        "native": "ðŸ˜€",
        "skin": null,
        "unified": "1f600",
      }
    }
    filter="bookface"
    index={0}
    key="0"
    onChange={[Function]}
  />
  <OverrideInput
    canDelete={false}
    emoji={
      Object {
        "colons": ":grinning:",
        "emoticons": Array [],
        "id": "grinning",
        "name": "Grinning Face",
        "native": "ðŸ˜€",
        "skin": null,
        "unified": "1f600",
      }
    }
    filter=""
    index={1}
    key="new"
    onChange={[Function]}
  />
</div>
`;
